_posts/2020-07-30-axon-kompleksowe-testowanie-aplikacji.md

```diff
@@ -0,0 +1,386 @@
+
+    @CommandHandler
+    public void handle(FetchTrailersCommand command) {
+        log.info("MOCK fetching...");
+        eventGateway.publish(new TrailersDetailsEvent(command.getProxyId(), TRAILERS));
+    }
+}
+```
+
+W teście chcę zasymulować sytuację, w której pojawia się command **CreateTrailersCommand**.
+W efekcie aplikacja powinna pobrać trailery dla wskazanego filmu i umieścić je w bazie:
+```java
+public class TrailersIntegrationTest extends CommonIntegrationSetup {
+    @Autowired
+    private CommandGateway commandGateway;
+    ...
+
+    @BeforeEach
+    public void beforeEach() {
+        this.movieId = "123";
+        this.trailers = generateTrailers(movieId);
+    }
+
+    @Test
+    public void shouldRetrieveTrailers() {
+        // given
+        commandGateway.send(new CreateTrailersCommand(
+                    trailers.getAggregateId(), 
+                    trailers.getExternalMovieId(), 
+                    trailers.getMovieId()));
+
+        await()
+            .atMost(FIVE_SECONDS)
+            .with()
+            .pollInterval(ONE_HUNDRED_MILLISECONDS)
+            .until(() -> trailerRepository.findByMovieId(movieId).isPresent());
+
+        // when
+        ResponseEntity<TrailerDTO[]> trailerResponse = testRestTemplate
+                .getForEntity(
+                        String.format(GET_TRAILERS_URL, randomServerPort, movieId),
+                        TrailerDTO[].class);
+        // then
+        assertThat(trailerResponse.getStatusCode()).isEqualTo(HttpStatus.OK);
+        assertThat(trailerResponse.getBody()).isNotNull();
+
+        List<TrailerDTO> trailerDTOS = Arrays.asList(trailerResponse.getBody());
+
+        assertThat(trailerDTOS.size()).isEqualTo(2);
+        assertThat(trailerDTOS).isEqualTo(trailers.getTrailers());
+    }
+}
+```
+
+Skorzystałem z [**awaitility**](https://github.com/awaitility/awaitility), żeby poczekać chwilę na odpowiedź w razie małej czkawki.  
+
+# Testy E2E
+Na samym szczycie piramidy testów są testy end-to-end, czyli sprawdzenie aplikacji w ten sposób, w jaki klient z niej korzysta.
+W moim przypadku klientem jest aplikacja frontendowa, która uderzając na konkretny endpoint, oczekuje konkretnej odpowiedzi.
+Przetestujmy wyszukanie filmu po tytule.
+
+Taki test powinien:
+- uderzyć na endpoint, za którym kryje się dana funkcjonalność,
+- sprawdzić status odpowiedzi od serwera,
+- sprawdzić zawartość odpowiedzi,
+- upewnić się, że film został umieszczony w bazie, a jeśli tak to czy jest on równy temu, co dostaliśmy w ciele odpowiedzi.
+
+```java
+public class CommonIntegrationSetup {
+    ...
+    protected ResponseEntity<MovieDTO> storeMovie(String title) {
+        ResponseEntity<MovieDTO> responseMovie = testRestTemplate.postForEntity(
+                        String.format(GET_OR_POST_MOVIES, randomServerPort), 
+                        new TitleBody(title), 
+                        MovieDTO.class);
+        moviesToCleanAfterTests.add(responseMovie.getBody());
```

Comment:
moviesToCleanAfterTests brzmi niebezpiecznie :D

Reply:
Z racji tego, że startuję jednego mongosa dla wszystkich test suite w ramach serwisu, to zaczęło mi to przeszkadzać w kilku testach, więc każdy teraz po sobie sprząta. :smile: 
Na potrzeby wpisu ta linia jest niepotrzebna więc ją usunę.